chips/pc/src/chip.rs

```diff
@@ -0,0 +1,235 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+use core::fmt::Write;
+use core::mem::MaybeUninit;
+
+use kernel::component::Component;
+use kernel::platform::chip::Chip;
+
+use tock_x86::mpu::PagingMPU;
+use tock_x86::support;
+use tock_x86::{Boundary, InterruptPoller};
+
+use crate::pit::{Pit, RELOAD_1KHZ};
+use crate::serial::{SerialPort, SerialPortComponent, COM1_BASE, COM2_BASE, COM3_BASE, COM4_BASE};
+
+use x86::bits32::paging::{PD, PT};
+
+/// Representation of a generic PC platform.
+///
+/// This struct serves as an implementation of Tock's [`Chip`] trait for the x86 PC platform. The
+/// behavior and set of peripherals available on PCs is very heavily standardized. As a result, this
+/// chip definition should be broadly compatible with most PC hardware.
+///
+/// Parameter `PR` is the PIT reload value. See [`Pit`] for more information.
+pub struct Pc<'a, const PR: u16 = RELOAD_1KHZ> {
+    /// Legacy COM1 serial port
+    pub com1: &'a SerialPort<'a>,
+
+    /// Legacy COM2 serial port
+    pub com2: &'a SerialPort<'a>,
+
+    /// Legacy COM3 serial port
+    pub com3: &'a SerialPort<'a>,
+
+    /// Legacy COM4 serial port
+    pub com4: &'a SerialPort<'a>,
+
+    /// Legacy PIT timer
+    pub pit: Pit<'a, PR>,
+
+    /// System call context
+    syscall: Boundary,
+    paging: PagingMPU<'a>,
+}
+
+impl<'a, const PR: u16> Chip for Pc<'a, PR> {
+    type MPU = PagingMPU<'a>;
+    fn mpu(&self) -> &Self::MPU {
+        &self.paging
+    }
+
+    type UserspaceKernelBoundary = Boundary;
+    fn userspace_kernel_boundary(&self) -> &Self::UserspaceKernelBoundary {
+        &self.syscall
+    }
+
+    fn service_pending_interrupts(&self) {
+        InterruptPoller::access(|poller| {
+            while let Some(num) = poller.next_pending() {
+                match num {
+                    32 => self.pit.handle_interrupt(),
```

Comment:
Could we define these interrupt numbers as constants?

Reply:
Const items added in latest push